test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,394 @@
+        LOGGER.error(exc.output.decode())
+        raise
+
+
+BACKUP_SCRIPT = textwrap.dedent(
+    """\
+#!/bin/bash -e
+
+DESTINATION_DIR=$PWD
+REMOTE_TMP_DIR=/tmp/zk_backup
+
+usage() {
+  echo "Usage: $0 [-h] [-p ZK_BACKUP_DEST_DIR] [-i SSH_KEY_PATH] [-t REMOTE_TMP_DIR] SSH_USER@MASTER_HOST" 1>&2
+}
+exit_abnormal() {
+  usage
+  exit 1
+}
+while getopts ":hp:t:i:" opt; do
+  case $opt in
+    h)
+      usage
+      exit 0
+    ;;
+    p) DESTINATION_DIR=$OPTARG
+    ;;
+    t) REMOTE_TMP_DIR=$OPTARG
+    ;;
+    i) SSH_KEY_PATH=$OPTARG
+    ;;
+    \\?) echo "Invalid option -$OPTARG" 1>&2
+        exit_abnormal
+    ;;
+    :) echo "Error: -$OPTARG requires an argument." 1>&2
+       exit_abnormal
+    ;;
+    *) exit_abnormal
+    ;;
+  esac
+done
+
+shift $((OPTIND - 1))
+if (($# != 1)); then
+  exit_abnormal
+fi
+
+CONN=$1
+KEY=${SSH_KEY_PATH:+"-i $SSH_KEY_PATH"}
+OPTS='-o ConnectTimeout=3 -o ConnectionAttempts=1
+-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no'
+
+ssh $OPTS $KEY -tt $CONN << EOF
+  set -e
+  sudo mkdir -p $REMOTE_TMP_DIR
+  sudo systemctl stop dcos-exhibitor
+  sudo cp -pr /var/lib/dcos/exhibitor/zookeeper ${REMOTE_TMP_DIR}/zookeeper
+  sudo systemctl start dcos-exhibitor
+  sudo tar --exclude 'myid' --exclude 'zookeeper.out' -pcvzf \\
+  ${REMOTE_TMP_DIR}/zk_backup-'$(date +%Y-%m-%d_%H-%M-%S)'.tar.gz -C $REMOTE_TMP_DIR ./zookeeper
+  sudo rm -rf ${REMOTE_TMP_DIR}/zookeeper
+  exit 0
+EOF
+
+scp $OPTS $KEY -r ${CONN}:${REMOTE_TMP_DIR}/zk_backup-'*'.tar.gz ${DESTINATION_DIR}/ 2> /dev/null
+
+ssh $OPTS $KEY -tt $CONN << EOF
+  set -e
+  sudo rm -rf $REMOTE_TMP_DIR
+  exit 0
+EOF
+exit 0
+    """
+)
+
+
+RESTORE_SCRIPT = textwrap.dedent(
+    """\
+#!/bin/bash -e
+
+SSH_USER='root'
+REMOTE_TMP_DIR=/tmp/zk_restore
+
+usage() {
+  echo "Usage: $0 [-h] [-i SSH_KEY_PATH] [-l SSH_USER] [-t REMOTE_TMP_DIR] ZK_BACKUP_PATH MASTER_HOSTNAME..." 1>&2
+}
+exit_abnormal() {
+  usage
+  exit 1
+}
+while getopts ":ht:i:l:" opt; do
+  case $opt in
+    h)
+      usage
+      exit 0
+    ;;
+    t) REMOTE_TMP_DIR=$OPTARG
+    ;;
+    i) SSH_KEY_PATH=$OPTARG
+    ;;
+    l) SSH_USER=$OPTARG
+    ;;
+    \\?) echo "Invalid option -$OPTARG" 1>&2
+        exit_abnormal
+    ;;
+    :) echo "Error: -$OPTARG requires an argument." 1>&2
+       exit_abnormal
+    ;;
+    *) exit_abnormal
+    ;;
+  esac
+done
+
+shift $((OPTIND - 1))
+if (($# < 2)); then
+  exit_abnormal
+fi
+
+BACKUP_PATH=$1
+shift
+
+OPTS='-o ConnectTimeout=3 -o ConnectionAttempts=1
+-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no'
+KEY=${SSH_KEY_PATH:+"-i $SSH_KEY_PATH"}
+
+for i in "$@"
+do
+ssh $OPTS -tt $KEY ${SSH_USER}@${i} << EOF
+  set -e
+  sudo systemctl stop dcos-exhibitor
+  sudo mkdir -p $REMOTE_TMP_DIR
+  exit 0
+EOF
+scp $OPTS $KEY $BACKUP_PATH $SSH_USER@${i}:${REMOTE_TMP_DIR}/zk_backup.tar.gz 2> /dev/null
+done
+
+for i in "$@"
+do
+ssh $OPTS -tt $KEY ${SSH_USER}@${i} << EOF
+  set -e
+  sudo rm -rf /var/lib/dcos/exhibitor/zookeeper
```

Comment:
Yes, but that undo is typically worth 18GB of data :)

Reply:
Also I haven't found a simple way to make the script idempotent then.
`cp` does a `stat` which fails on repeat.

Reply:
Maybe move this directory to `/var/lib/dcos/exhibitor/zookeeper.old` to provide some form of undo?

Reply:
done now.